src/databricks/sql/cloudfetch/download_manager.py

```diff
@@ -0,0 +1,139 @@
+                handler.result_file,
+                handler.result_link.startRowOffset,
+                handler.result_link.rowCount,
+            )
+            self.cloud_fetch_index += handler.result_link.rowCount
+            self.download_handlers.pop(idx)
+            # Return True upon successful download to continue loop and not force a retry
+            return result
+        # Download was not successful for next download item, force a retry
+        return None
+
+    def _remove_past_handlers(self, next_row_index):
+        """
+        Remove any download handlers whose start to end range doesn't include the next row to be fetched
+        i.e. no need to download
+        """
+        i = 0
+        while i < len(self.download_handlers):
+            result_link = self.download_handlers[i].result_link
+            if result_link.startRowOffset + result_link.rowCount > next_row_index:
+                i += 1
+                continue
+            self.download_handlers.pop(i)
+
+    def _schedule_downloads(self):
+        """
+        Schedule downloads for all download handlers if not already scheduled
+        """
+        for handler in self.download_handlers:
+            if handler.is_download_scheduled:
+                continue
+            try:
+                self.thread_pool.submit(handler.run)
+            except:
+                break
+            handler.is_download_scheduled = True
+
+    def _find_next_file_index(self, next_row_index):
+        # Get the next downloaded file
+        next_indices = [i for i, handler in enumerate(self.download_handlers)
+                        if handler.is_download_scheduled and handler.result_link.startRowOffset == next_row_index]
+        return next_indices[0] if len(next_indices) > 0 else None
+
+    def _check_if_download_successful(self, handler):
+        if not handler.is_file_download_successful():
+            if handler.is_link_expired:
+                self._stop_all_downloads_and_clear_handlers()
+                self.fetch_need_retry = True
+                return False
+            elif handler.is_download_timedout:
+                if self.num_consecutive_result_file_download_retries >= \
+                        self.downloadable_result_settings.max_consecutive_file_download_retries:
+                    # raise Exception("File download exceeded max retry limit")
+                    self.fetch_need_retry = True
+                    return False
+                self.num_consecutive_result_file_download_retries += 1
+                self.thread_pool.submit(handler)
+                return self._check_if_download_successful(handler)
+            else:
+                self.fetch_need_retry = True
+                return False
+
+        self.num_consecutive_result_file_download_retries = 0
+        self.fetch_need_retry = False
+        return True
+
+    def _stop_all_downloads_and_clear_handlers(self):
+        self.download_handlers = []
```

Comment:
Is this all? It seems like this will leave the ongoing threads running until a garbage collector runs. I expect to see a call to the ThreadPoolExecutor that actually kills those threads.

Reply:
I added `thread_pool.shutdown()`